### PR TITLE
Fix constrained tabbing failures with  Safari and Firefox

### LINKED ISSUE #41500: Can not scroll keyboard shorcut window with cursor
### Description

When opening keyboard shortcut window, it is not possible to navigate with the cursor to review all the keyboard shortcuts.

### Step-by-step reproduction instructions

1. Go to Keyboard shortcuts window
2. Use Tab to and cursor to navigate
3. Observe the issue: no option to see all shortcuts

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 6.0
- Browser: Chrome 102.0.5005.61

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #46041: Constrained tabbing: more failures with Safari and Firefox
### Description

Related: https://github.com/WordPress/gutenberg/pull/46038 and #45809
See also https://github.com/WordPress/gutenberg/pull/42653

Migrating the `a11y` tests to Playwright https://github.com/WordPress/gutenberg/pull/46038 surfaced more failures of Constrained tabbing, specifically with Safari and Firefox. Actually, in https://github.com/WordPress/gutenberg/pull/46038 one test fails with Webkit and two tests fail with Firefox. These can be reproduced also manually.

- The failure with Safari is an actual regression, as the specific test case worked well before https://github.com/WordPress/gutenberg/pull/34836
- The failures with Firefox isn't a regression but the actual behavior changed anyways after #34836.

### Step-by-step reproduction instructions

To manually reproduce the failing tests in the Draft PR https://github.com/WordPress/gutenberg/pull/46038

- Include the changes from https://github.com/WordPress/gutenberg/pull/42653 if not merged yet. Thie will fix another bug in Firefox.
- Edit a post.
- Open the Keyboard shortcuts modal dialog, for example by pressing Control+Option+H on macOS or by clicking the related item in the Options menu.
- Note: this modal dialog has two features that have an impact on the Constrained tabbing failure:
  - For all browsers other than Firefox: the modal contains only one focusable element: the Close button.
  - For Firefox: the modal contains a scrollable div element. Firefox makes scrollable element focusable when they trigger an actual overflow (your screen height may have an impact here, make sure the modal content height actually exceeds the modal container height and the div is actually scrollable).

**Safari:**
- With the modal opened, click any part of the modal after the Close button.
- Press the Tab key.
- The modal closes.
- Expected: the modal to stay opened and focus to move to the Close button.

This happens only when clicking somewhere _after_ the Close button.
- With the modal opened, click any part of the modal_before_ the Close button, for example click the modal title 'Keyboard shortcuts'.
- Press the Tab key.
- Observe the modal stays opened and focus goes to the Close button, as expected.

Firefox:
- Make sure the modal content is actually scrollable. This may depend on your screen size. When there is an actual overflow of the content, Firefox makes the scrollable div with CSS `overflow: auto` focusable.
- With the modal opened, press the Tab key. 
- Focus moves to the scrollable div, as Constrained tabbing relies on native browser behavior.
- Note that you won't see any focus style. To make debugging easier, add the following CSS somewhere:

```
:focus {
	outline: 4px solid red !important;
	outline-offset: -4px;
}
```

- Press the Tab key again.
- ~Focus stays on the scrollable div. The Close button is not reachable by pressing the Tab key.~ Edit: can't reproduce this part any longer.
- ~Expected: focus to go to the Close button and then cycle through the two elements at following Tab key presses.~
- Close the modal and then reopen it.
- Press Shift+Tab.
- Focus goes to the Close button.
- Press Shift+Tab again.
- Focus stays on the Close button. The scrollable div is not reachable by pressing the Shift+Tab.
- Expected: focus to go to the scrollable div and then cycle through the two elements at following Shift+Tab keys presses.

Firefox is the only browser that natively makes scrollable div elements focusable. I'm not sure this has been ever taken into account. However, the current implementation of Constrained tabbing relies on native browsers behaviors which are actually different. It seems to me the current implementation is inherently fragile as it is exposed to different, legitimate, browsers features. I'm not sure what is the best option here:
- either keep the current implementation and then cover all the browsers differences.
- revert to the previous implementation, which didn't rely on native browser behaviors and kept things more under control.

Either ways, the scrollable div issue would need to be solved by standardizing the behavor. There are good accessibility reasons to make a scrollable div always focusable, see #45809.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/46041
Fixes https://github.com/WordPress/gutenberg/issues/41500

## What?

The refactoring of `useConstrainedTabbing` in https://github.com/WordPress/gutenberg/pull/34836 introduced a few regressions and unexpected behaviors. Some of these regressions have been fixed already, see for example https://github.com/WordPress/gutenberg/issues/42652, https://github.com/WordPress/gutenberg/issues/45903, ,and https://github.com/WordPress/gutenberg/pull/42653. 

This PR aims to fix one remaining regression with Safari and some unexpected behavior with Firefox.

## TL;DR
The regression and unexpected behaviors surfaced only when attempting to run the existing [editor/various/a11y.spec.js](https://github.com/WordPress/gutenberg/blob/98ce5c53c9492764aeff88852a4073fb32e784ac/test/e2e/specs/editor/various/a11y.spec.js) test with all browsers in https://github.com/WordPress/gutenberg/pull/46038. The outcome was 1 failure with  webkit (Safari) and 2 failures with Firefox.

That proved the native browsers behavior differs in some cases. Therefore, the current implementation of `useConstrainedTabbing`, which relies on native browser behavior, doesn't appear to be ideal in the first place.

I'd like to note that when it comes to some accessibility features, it is very, very important to run E2E tests with all ajor browsers. In fact, it turned out browser behavior differs in at least two cases:
- Safari: when clicking somewhere after the last tabbable within a modal dialog and then pressing Tab, the modal dialog closes unexpectedly. Instead, wIth Chrome and Firefox the tab sequence restarts from the previous tababble. This seems related to the so called [Sequential focus navigation starting point](https://chromestatus.com/feature/5671747802103808), which is a browser feature that is apparently implemented differently in Safari.
- Firefox: it's the only major browser that natively makes scrollable elements focusable. The `useConstrainedTabbing` current implementation allows to tab to scrollable elements but it doesn't take them into account as 'tabbable' elements, leading to unexpected behavior.

I'd also like to kindly note that it appears the refactoring in https://github.com/WordPress/gutenberg/pull/34836 didn[t introduce any new test for the specific use case it was meant to address, see https://github.com/WordPress/gutenberg/issues/34681. I will create a separate issue for that.

## Why?

The constrained tabbing feature is broken since more than one year now. It needs to work reliably and consistently across all major browsers. It also needs to be tested more effectively.

## How?

- Adds some actual focus management when a tabbable element contains the next tabbable element. In this case, Safari seems to disagree with other browsers when there is no focused element and the 'Sequential focus navigation starting point' feature comes into play.
- Determines whether the modal dialog content is scrollable, that is: when there is an actual content overflow.
- Makes the scrollable div focusable, only when there is an actual content overflow. This reproduces the native Firefox behavior for all browsers.
- Use `ResizeObserver` to listen to the modal dialog content size changes and checks again whether it's scrollable. This is necessary because:
  - When resizing browser window, zooming in/out, changing orientation on mobile devices, etc., the content may overflow or not depending on the actual size of the content area.
  - Sometimes the modal dialog content is dynamic so that its size changes on the fly, e.g. the Preferences modal uses tabs. Some tabs produce a content overflow, some don't.
- Wraps the dialog content within a `<div>` element: this is necessary to determine size changes via `ResizeObserver`.
- Adds `aria-label="Scrollable section" tabindex="0"` to the dialog content when it's scrollable 
- Adds a `scrollableContentLabel` prop that can be used to override the default label `Scrollable section`.
- Adds a basic focus style to the content when it's scrollable and focusable.
- Makes the a11y E2E test run also with Firefox and Webkit.
- Adds E2E test for the scrollable section.

Note:
- I opted to not use `useResizeObserver` for now. I'm not sure I like the current implementation which injects an additional `<div>` element and overlays it on the node to check for resizing. Also, that works only when the node to check for is a positioned element. 
- It would be great to abstract and futher improve the scrollable section functionality in https://github.com/WordPress/gutenberg/issues/45809. Before that, it would be great to wait for https://github.com/WordPress/gutenberg/pull/41001 which aims to refactor `useResizeObserver` so that it can be used directly on an element. /Cc @youknowriad 

## Testing Instructions

- Run `npm run test:e2e:playwright -- --headed editor/various/a11y.spec.js` and check the tests pass.
- Edit a post and open the Keyboard shortcuts modal dialog.
- Tab through the elements within the modal dialog and check the list of shortcuts container receives focus and shows a focus style.
- With the scrollable element focused, check you can scroll the element by using the Down and Up arrow keys or the Spacebar / Shift+Spacebar. With Safari you may need to press Fn+Down/Up arrow.
- Check you can cycle through the Close button and the scrollable element by pressing Tab and Shift+Tab.
- Close the Keyboard shortcuts modal dialog and open the Preferences one.
- Depending also on your screen size, the General tab should not produce content overflow so it should not be scrollable.
- Inspect the source and check the content area, that is the element with `role=document` does _not_ have an `aria-label` attribute nor a `tabindex` attribute.
- Click on the Blocks tab. This tab has longer content and, depending also on your screen size, it should be scrollable.
- Check the content area is now scrollable and does have `aria-label="Scrollable section" tabindex="0"` attributes.
- Type `qwerty` in the 'Search for a block' search field.
- The modal dialog content will be shorter and only display 'No blocks found'.
- Check the content area is no longer scrollable nor focusable.
- Check it does _not_ have an `aria-label` attribute nor a `tabindex` attribute.
- Clear the search field. The content will show the long list of blocks again.
- Check the content area is scrollable and focusable again.
- Check it does have `aria-label="Scrollable section" tabindex="0"` attributes.
- Check there are no visual regressions in other modal dialogs e.g.:
  - the 'Explore all patterns' modal dialog
  - the confirm dialog when switching back to a draft
  - the confirm dialog when changing post visibility to 'Private'
  - the 'Lock block' modal dialog
  - the Query loop block > 'Choose a pattern' modal dialog
  - any other complex modal dialog e.g. in the Site Editor

### Testing Instructions for Keyboard

See above.

### CODE REVIEW

**Comment by youknowriad** [comment]
Hi @afercia it looks like this PR contains two separate fixes. One for constrained tabbing and one for modals? Does it make sense to split these into two smaller PRs? 

While the change for the constrained tabbing behavior seems quiet straightforward, I'd like to understand more the changes for the modal component.

**Comment by afercia** [comment]
Hi @youknowriad Sure I might split this fix into two smaller PRs but I'm not sure it would make much sense. While this PR touches two components, it actually fixes _one_ feature: constraining tabbing.
If I split, the a11y tests will still fail, either with Safari or Firefox.
If you really have a strong preference for splitting, I have no objections though.
Glad to help clarify the changes, if needed.

**Comment by afercia** [comment]
Tests keep failing for unrelated reasons, mostly timeouts or crashes e.g. `page.reload: Navigation failed because page crashed!`. Super annoying. Will rebase, not that I hope that will solve anything.

**Comment by youknowriad** [comment]
Sorry for the delay. If these are part of the same issue, it's fine to keep them together.

I have some questions here:

 - The "scrollable" related fix seems to only be applied to `Modal` component. What happens to the other places where we use `useConstrainedTabbing` (Thinking about popovers, ...). Isn't it better to apply that fix as a hook within `useDialog` or something like that, so that way it applies everywhere?

 - Why did you chose to apply "Firefox"'s behavior to all the other browsers instead of doing the opposite? Was that the simplest option?

**Comment by afercia** [comment]
@youknowriad thanks for your feedback. I'll try to do my best to answer your question.

The 'scrollable' fix will need to be applied wherever there is a scrollable element. That's not limited to the Modal or popover components. So I'm not sure `useDialog` would be a good place for this kind of fix. I plan to improve the existing `Scrollable` component in https://github.com/WordPress/gutenberg/issues/45809 and incorporate the accessibility improvements there. Once that is done, the improved `Scrollable` component should be used within the Modal to replace this temporary fix and everywhere there is a scrollable element. For now, this PR aims to fix the 'scrollable' thing within the Modal because it breaks constrained tabbing. 

> Why did you chose to apply "Firefox"'s behavior to all the other browsers instead of doing the opposite?

Scrollable elements need to be keyboard accessible to solve cases like https://github.com/WordPress/gutenberg/issues/41500. To do that, the scrollable element needs to be focusable in the first place. Firefox is the only browser that natively makes scrollable elements focusable. It does that only when there's an actual content overflow. You can test native browsers behavior [on this codepen](https://codepen.io/afercia/full/Expxgby).
Other browsers don't make scrollable elements focusable so they're not keyboard accessible. With these browsers it's not possible to scroll a scrollable element with a keyboard, unless it contains focusable elements. [The recommended technique](https://www.tpgi.com/short-note-on-improving-usability-of-scrollable-regions/) standardizes the Firefox accessible behavior across all browsers. This PR makes sure that this technique is applied 'dynamically' as in: only when there's an actual content overflow. 'Doing the opposite' wold mean making all scrollable elements not focusable in all browsers, so that they would be not operable with a keyboard in all browsers.

File: packages/components/src/modal/index.tsx [hunk 1]
```diff
@@ -172,10 +209,19 @@ function UnforwardedModal(
...
 						role="document"
 						onScroll={ onContentContainerScroll }
+						ref={ contentRef }
+						aria-label={
+							hasScrollableContent
+								? scrollableContentLabel ||
+								  __( 'Scrollable section' )
+								: undefined
+						}
+						tabIndex={ hasScrollableContent ? 0 : undefined }
```
Comment (youknowriad) [inline packages/components/src/modal/index.tsx:222]: Do you think we can extract all this logic into a hook `const { ref, props } = useScrollable();` 

and just basically apply the "ref" and "props" to the element that is supposed to be scrollable? I think it might bring some clarity to the code here.

Comment (youknowriad) [reply packages/components/src/modal/index.tsx:222]: Also, not sure why we need `childrenContainerRef` can't we just check the height of the scrollable itself (or at least its children) ?

**Comment by aristath** [comment]
This is a bugfix and as such, it looks good to me.
I agree with @youknowriad that it should be done in a hook. However, I think that perhaps the abstraction should probably be done in a follow-up PR? This way we can unblock this one, fix the bug, and continue with the abstraction separately.

File: packages/components/src/modal/index.tsx [hunk 2]
```diff
@@ -65,6 +67,7 @@ function UnforwardedModal(
 		onKeyDown,
 		isFullScreen = false,
 		__experimentalHideHeader = false,
+		scrollableContentLabel,
```
Comment (youknowriad) [inline packages/components/src/modal/index.tsx]: This prop is the only API change in this PR. I'm fine with this PR if we don't add maintenance burden for us like that. 

So in the sense, I suggested the refactor to improve the code quality a bit and clarify how this works but I'm fine shipping this as is if we only focus on the bug fix without new APIs...

Comment (afercia) [reply packages/components/src/modal/index.tsx]: Ok fair enough. Will remove it and keep the default `aria-label={ __( 'Scrollable section' ) }`

File: packages/components/src/modal/index.tsx (hunk 1 above)
Comment (afercia) [reply packages/components/src/modal/index.tsx:222]: The scrollable element has a fixed height,
The height of the content (the children) determines the actual content overflow, when it's greater than the scrollable height.
There may be multipole children though. Not sure how to calculate eh height of multiple children that could also be dynamically rendered. Seems to me the simplest option is to check for the height of an additional container.

Comment (afercia) [reply packages/components/src/modal/index.tsx:222]: > Do you think we can extract all this logic into a hook 

I'm all for better abstraction and improvements. I'd prefer to do that when enhancing the `Scrollable` component though. See https://github.com/WordPress/gutenberg/issues/45809

**Review by youknowriad — APPROVED** [review]
code is good to ship 👍 Might be good to get another a11y check though. @alexstine

File: test/e2e/specs/editor/various/a11y.spec.js [hunk 3]
```diff
@@ -93,4 +111,87 @@ test.describe( 'a11y', () => {
...
+		await expect( preferencesModal ).toBeFocused();
+
+		// Check the General tab panel is visible by default.
+		await expect(
+			preferencesModal.locator( 'role=tabpanel[name="General"i]' )
+		).toBeVisible();
+
+		async function clickAndFocusTab( tab ) {
+			await tab.click();
+			await tab.focus();
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/a11y.spec.js:159]: Why is this necessary? Wouldn't clicking automatically focus the tab?

**Review by kevin940726 — COMMENTED** [review]
Test looks good 👍 . Just a small question there.

File: test/e2e/specs/editor/various/a11y.spec.js (hunk 3 above)
Comment (youknowriad) [reply test/e2e/specs/editor/various/a11y.spec.js:159]: > Wouldn't clicking automatically focus the tab?

Not in all browsers, this behavior depends on the browser and the OS and it's actually the source of so many a11y bugs.

Comment (kevin940726) [reply test/e2e/specs/editor/various/a11y.spec.js:159]: I see, TIL! Then I think it's fine as long as it's not part of the user flow we want to test 👍 .

Comment (afercia) [reply test/e2e/specs/editor/various/a11y.spec.js:159]: Safari 😞 
I'll clarify with a comment.

Comment (afercia) [reply test/e2e/specs/editor/various/a11y.spec.js:159]: Safari (*webkit) 😞 
I'll clarify with a comment.

**Comment by afercia** [comment]
Thanks everyone for the feedback and review.
Happy to see the tests pass with chromium, firefox, and webkit.
Everything is green but I'll wait a couple days to see if @alexstine has time for one more a11y check.

I commented on https://github.com/WordPress/gutenberg/issues/45809 reporting the findings from this PR and the suggestion to craft a new hook. It will need to wait for https://github.com/WordPress/gutenberg/pull/41001 anyways.

**Comment by alexstine** [comment]
I do not think I can really add much to this PR. I trust your judgement @afercia . Thanks.

**Comment by afercia** [comment]
Thanks. Merging 🎉
